File: src/org/cappuccino_project/ide/intellij/plugin/stubs/types/ObjJAbstractFunctionDeclarationStubType.java
package org.cappuccino_project.ide.intellij.plugin.stubs.types;

import com.intellij.lang.ASTNode;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.psi.stubs.IndexSink;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import com.intellij.util.io.StringRef;
import org.cappuccino_project.ide.intellij.plugin.indices.StubIndexService;
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJFunctionDeclarationElement;
import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.ObjJFunctionDeclarationElementStub;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public abstract class ObjJAbstractFunctionDeclarationStubType<PsiT extends ObjJFunctionDeclarationElement<? extends ObjJFunctionDeclarationElement>, Stub extends ObjJFunctionDeclarationElementStub<PsiT>> extends ObjJStubElementType<ObjJFunctionDeclarationElementStub<PsiT>, PsiT> {
    ObjJAbstractFunctionDeclarationStubType(
            @NotNull
                    String debugName, Class<PsiT> functionDecClass, Class<Stub> stubClass) {
        super(debugName, functionDecClass, stubClass);
    }

    @NotNull
    @Override
    public ObjJFunctionDeclarationElementStub<PsiT> createStub(
            @NotNull
                    PsiT element, StubElement stubParent) {
        final String fileName = element.getContainingFile() != null && element.getContainingFile().getVirtualFile() != null ? element.getContainingFile().getVirtualFile().getName() : "undefined";
        final String functionNameString = element.getFunctionNameAsString();
        final List<String> paramNames = element.getParamNames();
        final String returnType = element.getReturnType();
        return createStub(stubParent, fileName, functionNameString, paramNames, returnType);
    }

    @NotNull
    abstract ObjJFunctionDeclarationElementStub<PsiT> createStub(StubElement parent,
                                                                 @NotNull
                                                                         String fileName,
                                                                 @NotNull
                                                                         String fqName,
                                                                 @NotNull
                                                                         List<String> paramNames,
                                                                 @Nullable
                                                                         String returnType);

    @Override
    public void serialize(
            @NotNull
                    ObjJFunctionDeclarationElementStub stub,
            @NotNull
                    StubOutputStream stream) throws IOException {

        stream.writeName(stub.getFileName());
        stream.writeName(stub.getFqName());
        stream.writeInt(stub.getNumParams());
        for (Object param : stub.getParamNames()) {
            stream.writeName((String)param);
        }
        stream.writeName(stub.getReturnType());
    }

    @NotNull
    @Override
    public ObjJFunctionDeclarationElementStub<PsiT> deserialize(
            @NotNull
                    StubInputStream stream, StubElement stubParent) throws IOException {
        final String fileName = StringRef.toString(stream.readName());
        final String fqName = StringRef.toString(stream.readName());
        final int numParams = stream.readInt();
        final List<String> paramNames = new ArrayList<>();
        for (int i=0;i<numParams;i++) {
            paramNames.add(StringRef.toString(stream.readName()));
        }
        final String returnType = StringRef.toString(stream.readName());
        return createStub(stubParent, fileName, fqName, paramNames, returnType);
    }

    @Override
    public void indexStub(@NotNull ObjJFunctionDeclarationElementStub stub, @NotNull IndexSink indexSink) {
        ServiceManager.getService(StubIndexService.class).indexFunctionDeclaration(stub, indexSink);
    }

    @Override
    public boolean shouldCreateStub(ASTNode node) {
        return node.getPsi() instanceof ObjJFunctionDeclarationElement;
    }
}

File: src/org/cappuccino_project/ide/intellij/plugin/stubs/types/ObjJAbstractFunctionDeclarationStubType.kt
package org.cappuccino_project.ide.intellij.plugin.stubs.types

import com.intellij.lang.ASTNode
import com.intellij.openapi.components.ServiceManager
import com.intellij.psi.stubs.IndexSink
import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import com.intellij.util.io.StringRef
import org.cappuccino_project.ide.intellij.plugin.indices.StubIndexService
import org.cappuccino_project.ide.intellij.plugin.psi.interfaces.ObjJFunctionDeclarationElement
import org.cappuccino_project.ide.intellij.plugin.stubs.interfaces.ObjJFunctionDeclarationElementStub

import java.io.IOException
import java.util.ArrayList

abstract class ObjJAbstractFunctionDeclarationStubType<PsiT : ObjJFunctionDeclarationElement<out ObjJFunctionDeclarationElementStub<*>>, Stub : ObjJFunctionDeclarationElementStub<PsiT>> internal constructor(
        debugName: String, functionDecClass: Class<PsiT>, stubClass: Class<Stub>) : ObjJStubElementType<ObjJFunctionDeclarationElementStub<PsiT>, PsiT>(debugName, functionDecClass, stubClass) {

    override fun createStub(
            element: PsiT, stubParent: StubElement<*>): ObjJFunctionDeclarationElementStub<PsiT> {
        val fileName = if (element.containingFile != null && element.containingFile.virtualFile != null) element.containingFile.virtualFile.name else "undefined"
        val functionNameString = element.functionNameAsString
        val paramNames = element.paramNames
        val returnType = element.returnType
        val shouldResolve = element.shouldResolve()
        return createStub(stubParent, fileName, functionNameString, paramNames, returnType, shouldResolve)
    }

    internal abstract fun createStub(parent: StubElement<*>,
                                     fileName: String,
                                     fqName: String,
                                     paramNames: List<String>,
                                     returnType: String?,
                                     shouldResolve: Boolean): ObjJFunctionDeclarationElementStub<PsiT>

    @Throws(IOException::class)
    override fun serialize(
            stub: ObjJFunctionDeclarationElementStub<*>,
            stream: StubOutputStream) {

        stream.writeName(stub.fileName)
        stream.writeName(stub.fqName)
        stream.writeInt(stub.numParams)
        for (param in stub.paramNames) {
            stream.writeName(param as String)
        }
        stream.writeName(stub.returnType)
        stream.writeBoolean(stub.shouldResolve())
    }

    @Throws(IOException::class)
    override fun deserialize(
            stream: StubInputStream, stubParent: StubElement<*>): ObjJFunctionDeclarationElementStub<PsiT> {
        val fileName = StringRef.toString(stream.readName())
        val fqName = StringRef.toString(stream.readName())
        val numParams = stream.readInt()
        val paramNames = ArrayList<String>()
        for (i in 0 until numParams) {
            paramNames.add(StringRef.toString(stream.readName()))
        }
        val returnType = StringRef.toString(stream.readName())
        val shouldResolve = stream.readBoolean()
        return createStub(stubParent, fileName, fqName, paramNames, returnType, shouldResolve)
    }

    override fun indexStub(stub: ObjJFunctionDeclarationElementStub<*>, indexSink: IndexSink) {
        ServiceManager.getService(StubIndexService::class.java).indexFunctionDeclaration(stub, indexSink)
    }

    override fun shouldCreateStub(node: ASTNode?): Boolean {
        return node!!.psi is ObjJFunctionDeclarationElement<*>
    }
}
